master/internal/sproto/globals.go

```diff
@@ -0,0 +1,42 @@
+package sproto
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/determined-ai/determined/master/pkg/actor"
+)
+
+var globalRM *actor.Ref
```

Comment:
I don't think using global variables is a good idea - it contradicts the decoupling of the actor system and will make it hard to turn this into a distributed system eventually.

Reply:
Why does it contract the decoupling of the actor system? It is still decoupled from the actual application. And I don't quite understand why you want to turn the actor system into a distributed system.

Reply:
The actor system uses message-passing to ensure decoupling of actors and adding in mutable shared state like this breaks the guarantees that the actor system gives us (i.e. allowing us to confidently write code using a single-threaded paradigm because each variable is only used by a single actor. Now this variable can be accessed and altered by different actors and we need to think about making this code safe for concurrency).

I imagine we are eventually going to want to spread the master across multiple machines since that is one of the major advantages of using an actor system. But that is probably debatable.

Reply:
> The actor system uses message-passing to ensure decoupling of actors and adding in mutable shared state like this breaks the guarantees that the actor system gives us (i.e. allowing us to confidently write code using a single-threaded paradigm because each variable is only used by a single actor. Now this variable can be accessed and altered by different actors and we need to think about making this code safe for concurrency).

Yes, you are right.  How about adding a lock for it? The problem with passing the reference of actors around is that sometimes there are mutual reference between actors, which are in this PR the references between the resource manager and the agents actor and between the resource manager and the task actor. So we have to use either dependency injection or find the dynamic reference finding `Ref.System().Get(...)`. The problem with dynamic reference is it depends on a variable string rather than a constant reference. It's pretty common to use a singleton pattern in such cases. In Golang, global functions are the way to implement it.

> I imagine we are eventually going to want to spread the master across multiple machines since that is one of the major advantages of using an actor system. But that is probably debatable.

That's very far from where we are right now. And even if we are going to build multi masters it doesn't necessarily have to use a distributed actor system nor use the current `system` struct to represent a distributed actor system like an RPC system.

Reply:
Talked offline. Will not use global variables to store the actor reference.